### PR TITLE
Added torchrun compatibility for distributet training across multiple GPUs in a single node (single instance)

### PR DESCRIPTION
**Pull request by brunopistone**
*Issue #, if available:*

*Description of changes:*

Added possibility to execute remote function with torchrun command, for parallelizing training across multiple GPUs in a single node (single instance).
The functionality can be enabled in the @remote function as following:

```
@remote(use_torchrun=True, nproc_per_node=2)
def train(....):
   pass

model = train(....)
```

*Testing done:*

Integration tests for evaluating back compatibility with original way of working, added integration test for checking compatibility with new functionality added

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by brunopistone** [comment]
Any update @mohanasudhan?

**Comment by mohanasudhan** [comment]
@mufaddal-rohawala  can you help with the review?

File: src/sagemaker/remote_function/client.py [hunk 1]
```diff
@@ -90,6 +88,8 @@ def remote(
     spark_config: SparkConfig = None,
     use_spot_instances=False,
     max_wait_time_in_seconds=None,
+    use_torchrun=False,
+    nproc_per_node=1,
```
Comment (mufaddal-rohawala) [inline src/sagemaker/remote_function/client.py:92]: We need to add docstrings for these params for better user experience? Also nit: `nproc_per_node ` might not be an apt name here, wanted to check if there is value to rename it to something more informative.

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]
Adding docstrngs is a blocker.

File: src/sagemaker/remote_function/client.py (hunk 1 above)
Comment (brunopistone) [reply src/sagemaker/remote_function/client.py:92]: Added docstring for `use_torchrun` and `nproc_per_node` under sagemaker/remote_function/client.py

**Review by sage-maker — DISMISSED** [review]

**Comment by sage-maker** [comment]
@brunopistone Please review the test failures, thanks

**Comment by sage-maker** [comment]
@brunopistone 
Looks like codestyle is still failing

File: src/sagemaker/remote_function/job.py [hunk 2]
```diff
@@ -951,7 +1001,12 @@ def _get_job_name(job_settings, func):
 
 
 def _prepare_and_upload_runtime_scripts(
-    spark_config: SparkConfig, s3_base_uri: str, s3_kms_key: str, sagemaker_session: Session
+        spark_config: SparkConfig,
```
Comment (sage-maker) [inline src/sagemaker/remote_function/job.py:1004]: Too much whitespace, looks like two tab instead of 1 maybe

**Review by sage-maker — CHANGES_REQUESTED** [review]

File: tests/integ/sagemaker/remote_function/test_decorator.py [hunk 3]
```diff
@@ -818,3 +818,25 @@ def test_decorator_auto_capture(sagemaker_session, auto_capture_test_container):
         f"--rm {auto_capture_test_container}"
     )
     subprocess.check_output(shlex.split(cmd), stderr=subprocess.STDOUT).decode("utf-8")
+
+def test_decorator_torchrun(
+        sagemaker_session,
```
Comment (sage-maker) [inline tests/integ/sagemaker/remote_function/test_decorator.py:823]: Same here, needs 1 tab instead of 2.

**Review by sage-maker — CHANGES_REQUESTED** [review]

**Review by sage-maker — APPROVED** [review]

**Comment by sage-maker** [comment]
@brunopistone 
Your codestyle still failed, refer to `client.py` line 60 and `test_decorator.py` line 822. Please make sure you are running the linters locally, thanks

File: src/sagemaker/remote_function/client.py [hunk 4]
```diff
@@ -58,7 +58,6 @@
 
 logger = logging_config.get_logger()
```
Comment (sage-maker) [inline src/sagemaker/remote_function/client.py:60]: Needs a new line here

File: tests/integ/sagemaker/remote_function/test_decorator.py [hunk 5]
```diff
@@ -818,3 +818,25 @@ def test_decorator_auto_capture(sagemaker_session, auto_capture_test_container):
         f"--rm {auto_capture_test_container}"
     )
     subprocess.check_output(shlex.split(cmd), stderr=subprocess.STDOUT).decode("utf-8")
+
```
Comment (sage-maker) [inline tests/integ/sagemaker/remote_function/test_decorator.py:821]: and a new line here

**Review by sage-maker — CHANGES_REQUESTED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

File: src/sagemaker/remote_function/job.py [hunk 6]
```diff
@@ -162,6 +162,51 @@
...
+    conda_env=$(cat remote_function_conda_env.txt)
+
+    if which mamba >/dev/null; then
+        conda_exe="mamba"
+    else
+        conda_exe="conda"
+    fi
+
+    printf "INFO: Invoking remote function with torchrun inside conda environment: $conda_env.\\n"
+    $conda_exe run -n $conda_env torchrun --nproc_per_node $NPROC_PER_NODE -m sagemaker.remote_function.invoke_function "$@"
```
Comment (sage-maker) [inline src/sagemaker/remote_function/job.py:203]: Line too long

**Review by sage-maker — APPROVED** [review]

**Pull request merged**